// hsf/lib/provider/m_auto_checker.js
/*!
 * hsf - provider/m_auto_check.js
 * Copyright(c) 2012 - 2013 Taobao.com
 * Author: busi.hyy <busi.hyy@taobao.com>
 */

/**
 * Module dependencies.
 */
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var AutoChecker = require('./auto_checker');

function MAutoChecker(options) {
  this.optionsList = [];
  this.checkers = [];
  EventEmitter.call(this);

  for (var i = 0; i < options.serviceName.length; i++) {
    var opt = {};
    for (var key in options) {
      if (key === '_id') {
        continue;
      }
      opt[key] = options[key];
    }
    opt._id = options.serviceName[i] + ':' + options.version[i];
    this.optionsList.push(opt);
    var checker = AutoChecker.create(opt);
    this.checkers.push(checker);
    checker.on('autocheck', this.emit.bind(this, 'autocheck'));
  }
};
util.inherits(MAutoChecker, EventEmitter);

MAutoChecker.prototype.registFunc = function () {
  for (var i = 0; i < this.checkers.length; i++) {
    this.checkers[i].registFunc.apply(this.checkers[i], arguments);
  }
};

MAutoChecker.prototype.start = function () {
  for (var i = 0; i < this.checkers.length; i++) {
    this.checkers[i].start();
  }
};

MAutoChecker.prototype.stop = function() {
  for (var i = 0; i < this.checkers.length; i++) {
    this.checkers[i].stop();
  }
};

exports.create = function (options) {
  return new MAutoChecker(options);
};
