// hsf/lib/provider/m_publisher.js
/*!
 * hsf - provider/m_publisher.js
 * Copyright(c) 2012 - 2013 Alibaba Group Holding Limited.
 * Authors: 
 *   busi.hyy <busi.hyy@taobao.com>

 */

var Publisher = require('./publisher');
var EventEmitter = require('events').EventEmitter;
var util = require('util');

var MPublisher = function (options) {
  this.optionsList = [];
  this.publishers = [];
  EventEmitter.call(this);

  for (var i = 0; i < options.serviceName.length; i++) {
    var opt = {};
    for (var key in options) {
      if (key === '_id') {
        continue;
      }
      opt[key] = options[key];
    }
    opt._id = options.serviceName[i] + ':' + options.version[i];
    this.optionsList.push(opt);
    this.publishers.push(Publisher.createPublisher(opt));
  }
  this.bindEvent();
};

util.inherits(MPublisher, EventEmitter);

MPublisher.prototype.bindEvent = function () {
  for (var i = 0; i < this.publishers.length; i++) {
    var publisher = this.publishers[i];
    publisher.on('connection', this.emit.bind(this, 'connection'));
    publisher.on('CSdisconnect', this.emit.bind(this, 'CSdisconnect'));
    publisher.on('error', this.emit.bind(this, 'error'));
    publisher.on('init', this.emit.bind(this, 'init'));
    publisher.on('refresh', this.emit.bind(this, 'refresh'));
    publisher.on('CSconnect', this.emit.bind(this, 'CSconnect'));
  }
};

MPublisher.prototype.publish = function (apiMeta) {
  for (var i = 0; i < this.publishers.length; i++) {
    this.publishers[i].publish(apiMeta);
  }
};

MPublisher.prototype.unPublish = function () {
  for (var i = 0; i < this.publishers.length; i++) {
    this.publishers[i].unPublish();
  }
};

exports.createPublisher = function (options) {
  return new MPublisher(options);
};
