// hsf/lib/index.js
/*!
 * hsf - index.js
 * Copyright(c) 2012 Taobao.com
 * Author: busi.hyy <busi.hyy@taobao.com>, suqian.yf <suqian.yf@taobao.com>
 */

var fs = require('fs');
var path = require('path');
var consumer = require('./consumer');
var provider = require('./provider');
fs.existsSync = fs.existsSync || path.existsSync;

exports.ConfigClient = require('./config_client');
exports.createClient = consumer.createClient;
exports.Consumer = consumer.Consumer;
exports.NoKeepaliveConsumer = consumer.NoKeepaliveConsumer;

exports.createServer = provider.createServer;
exports.HSFResponse = require('./provider/hsf_response');
exports.Publisher = require('./provider/publisher');

//config server
var configServer = require('./config_server/config_server');
configServer.Service = require('./config_server/service');
exports.configServer = configServer;

/**
 * Auto-load bundled middleware with getters.
 */

exports.middleware = {};
var middlewareDir = path.join(__dirname, 'middleware');

if (fs.existsSync(middlewareDir)) {
  fs.readdirSync(middlewareDir).forEach(function (filename) {
    if (!/\.js$/.test(filename)) {
      return;
    }
    var name = path.basename(filename, '.js');
    function load() {
      return require('./middleware/' + name);
    }
    exports.middleware.__defineGetter__(name, load);
    exports.__defineGetter__(name, load);
  });
}
