// hsf-protocol/lib/config_server.js
/**!
 * hsf-protocol - lib/hsf.js
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 *
 * Authors:
 *   busi.hyy <busi.hyy@taobao.com>
 */

'use strict';

/**
 * Module dependencies.
 */

var hessian = require('hessian.js');
var remoting = require('./remoting');
var java = require('js-to-java');

/**
 * generate configserver request body
 * @param {String} id
 * @param {String} group
 * @return {Buffer}
 */
exports.genReqBody = function (id, group) {
  var reqObj = java('com.taobao.config.server.common.dataobject.Command', {
    id: null,
    name: 'getGroupData',
    params: java('[object', [id, group])
  });

  return hessian.encode(reqObj);
};

/**
 * cs regist encode method
 * @param {String} clientId
 * @param {String} dataId
 * @param {String} group
 * @param {String} addr
 * @return {Buffer}
 */
exports.csRegistEncode = function (clientId, dataId, group, addr) {
  return java('com.taobao.config.common.protocol.PublisherRegReqPacket', {
    elements: [
      java('com.taobao.config.common.protocol.PublisherRegElement', {
        clientId: clientId,
        dataId: dataId,
        datumId: dataId + '@' + addr + '@hsf-nodejs'
      }),
      java('com.taobao.config.common.protocol.AttributeElement', {
        name: '!Group',
        value: group
      })
    ]
  });

};

/**
 * cs publish encode method
 * @param {String} clientId
 * @param {String} dataId
 * @param {String} data
 * @param {Number} revision
 * @return {Buffer}
 */
exports.csPublishEncode = function (clientId, dataId, data, revision) {
  data = new Buffer(data);
  var buf = new Buffer(5 + 2 + data.length);
  var head = new Buffer([0xAC, 0xED, 0x00, 0x05, 0x74]);
  head.copy(buf, 0, 0, 5);
  buf.writeInt16BE(data.length, 5, 0);
  data.copy(buf, 7, 0, data.length);

  return java('com.taobao.config.common.protocol.PublisherDataElement', {
    clientId: clientId,
    dataId: dataId,
    needAck: true,
    revision: java('com.taobao.config.common.Revision', {revision: java.long(revision)}),
    data: java('java.util.LinkedList', [
      java('com.taobao.config.common.protocol.Swizzle', {
        bytes: buf,
        serialization: 1
      })
    ])
  });
};

/**
 * cs do publish encode method
 * @param {Array} bufs
 * @return {Buffer}
 */
exports.csDoPublishEncode = function (bufs) {
  var params = {
    packetId: 0,
    data: bufs,
    type: 'cs'
  };
  return remoting.encode(params);
};
