// hsf-protocol/lib/constant.js
/**!
 * hsf-protocol - lib/constant.js
 * Copyright(c) 2014 Alibaba Group Holding Limited.
 *
 * Authors:
 *   busi.hyy <busi.hyy@taobao.com>
 */

'use strict';

/**
 * Module dependencies.
 */

var hessian = require('hessian.js');
var java = require('js-to-java');

/**
 * ConnectionRequest
 */
exports.CONNECTION_REQUEST_CLASSNAME = 'com.taobao.remoting.impl.ConnectionRequest';
exports.CONNECTION_REQUEST_CTX = 'com.taobao.remoting.impl.ConnectionRequest$RequestContext';
exports.CONNECTION_REQUEST_CLASSNAME_SIZE = exports.CONNECTION_REQUEST_CLASSNAME.length;


/**
 * ConnectionResponse
 */
exports.CONNECTION_RESPONSE_CLASSNAME = 'com.taobao.remoting.impl.ConnectionResponse';
exports.CONNECTION_RESPONSE_CLASSNAME_SIZE = exports.CONNECTION_RESPONSE_CLASSNAME.length;
exports.CONNECTION_RESPONSE_CTX = 'com.taobao.remoting.impl.ConnectionResponse$ResponseContext';

/**
 * ConnectionHeartBeat
 */
exports.CONNECTION_HEART_BEAT_CLASSNAME = 'com.taobao.remoting.impl.ConnectionHeartBeat';
exports.CONNECTION_HEART_BEAT_CLASSNAME_BUF = new Buffer(exports.CONNECTION_HEART_BEAT_CLASSNAME);

/**
 * request and response object
 */
exports.HSF_REQUEST_CLASSNAME = 'com.taobao.hsf.domain.HSFRequest';
exports.HSF_REQUEST_CLASSNAME_BUF = new Buffer(exports.HSF_REQUEST_CLASSNAME);

exports.HSF_RESPONSE_CLASSNAME = 'com.taobao.hsf.domain.HSFResponse';
exports.HSF_RESPONSE_CLASSNAME_BUF = new Buffer(exports.HSF_RESPONSE_CLASSNAME);

/**
 * ConnectConfigServer
 */
exports.CS_PROTOCOL_CLASSNAME = 'com.taobao.config.common.protocol.ProtocolPackage';
exports.CS_PROTOCOL_CLASSNAME_BUF = new Buffer(exports.CS_PROTOCOL_CLASSNAME);

exports.DEFAULT_CLASSNAME = {
  boolean: 'boolean',
  int: 'int',
  long: 'long',
  double: 'double',
  date: 'java.util.Date',
  string: 'java.lang.String',
  byteArray: '[B',
  list: 'java.util.ArrayList',
  map: 'java.util.HashMap',
  exception: 'java.lang.RuntimeException',
  null: 'null'
};

exports.MAX_INT_32 = Math.pow(2, 31);

exports.REQUEST_CONN_TPL = hessian.encode({
  $class: exports.CONNECTION_REQUEST_CLASSNAME,
  $: {
    ctx: {
      $class: exports.CONNECTION_REQUEST_CTX,
      $: { id: java.long(0) }
    }
  }
});
